File: webapp-core/src/resolver/graph.ts
// Dependency Graph Builder
// Core resolution logic with security hooks

import semver from 'semver';
import { 
  PackageRef, 
  MegagateConfig,
  SecurityCheckResult,
  LockdownCheckResult 
} from '../types/index.js';

export interface DependencyNode {
  name: string;
  range: string;
  resolved?: ResolvedDependency;
  dependencies: Map<string, DependencyNode>;
  dependents: Set<string>;
  // Security flags
  securityCheck?: SecurityCheckResult;
  lockdownCheck?: LockdownCheckResult;
}

export interface Conflict {
  name: string;
  versions: Array<{ version: string; requiredBy: string[] }>;
  resolution?: 'hoist' | 'duplicate' | 'error';
}

export interface ResolvedDependency {
  name: string;
  version: string;
  integrity: string;
  resolved: string;
  size: number;
  dependencies?: Record<string, string>;
  optionalDependencies?: Record<string, string>;
  peerDependencies?: Record<string, string>;
  bin?: Record<string, string>;
  engines?: Record<string, string>;
  scripts?: Record<string, string>;
  publishTime?: string;
  approvedBuilds?: string[];
}

export class DependencyGraph {
  private nodes = new Map<string, DependencyNode>();
  private config: MegagateConfig;
  private securityEnabled: boolean = false;

  constructor(config: MegagateConfig) {
    this.config = config;
  }

  enableSecurity() {
    this.securityEnabled = true;
  }

  addNode(name: string, range: string): DependencyNode {
    const key = `${name}@${range}`;
    if (!this.nodes.has(key)) {
      this.nodes.set(key, {
        name,
        range,
        dependencies: new Map(),
        dependents: new Set(),
      });
    }
    return this.nodes.get(key)!;
  }

  getNode(name: string, version: string): DependencyNode | undefined {
    const key = `${name}@${version}`;
    return this.nodes.get(key);
  }

  addEdge(parent: DependencyNode, child: DependencyNode): void {
    parent.dependencies.set(child.name, child);
    child.dependents.add(parent.name);
  }

  detectConflicts(): Conflict[] {
    const conflicts: Conflict[] = [];
    const nameToVersions = new Map<string, Set<string>>();

    for (const node of this.nodes.values()) {
      if (!nameToVersions.has(node.name)) {
        nameToVersions.set(node.name, new Set());
      }
      nameToVersions.get(node.name)!.add(node.range);
    }

    for (const [name, ranges] of nameToVersions.entries()) {
      if (ranges.size > 1) {
        const versions = Array.from(ranges).map(range => {
          const node = this.findNodeByRange(name, range);
          return {
            version: node?.resolved?.version || range,
            requiredBy: Array.from(node?.dependents || []),
          };
        });
        conflicts.push({ name, versions });
      }
    }

    return conflicts;
  }

  private findNodeByRange(name: string, range: string): DependencyNode | undefined {
    const key = `${name}@${range}`;
    return this.nodes.get(key);
  }

  getResolutionOrder(): DependencyNode[] {
    // Topological sort
    const visited = new Set<string>();
    const order: DependencyNode[] = [];

    const visit = (node: DependencyNode) => {
      const key = `${node.name}@${node.resolved?.version || node.range}`;
      if (visited.has(key)) return;
      visited.add(key);

      for (const dep of node.dependencies.values()) {
        visit(dep);
      }
      order.push(node);
    };

    for (const node of this.nodes.values()) {
      if (node.dependents.size === 0) { // Root nodes (no dependents)
        visit(node);
      }
    }

    return order;
  }

  setSecurityResult(nodeName: string, version: string, result: SecurityCheckResult): void {
    const key = `${nodeName}@${version}`;
    const node = this.nodes.get(key);
    if (node) node.securityCheck = result;
  }

  setLockdownResult(nodeName: string, version: string, result: any): void {
    const key = `${nodeName}@${version}`;
    const node = this.nodes.get(key);
    if (node) node.lockdownCheck = result;
  }
}

File: crates/megagate-resolver/src/graph.rs
use megagate_types::package::ResolvedDependency;
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone)]
pub struct DependencyNode {
    pub name: String,
    pub range: String,
    pub resolved: Option<ResolvedDependency>,
    pub dependencies: HashMap<String, DependencyNode>,
    pub dependents: HashSet<String>,
}

impl DependencyNode {
    pub fn key(&self) -> String {
        format!("{}@{}", self.name, self.range)
    }
}

#[derive(Debug, Clone)]
pub struct DependencyGraph {
    pub nodes: HashMap<String, DependencyNode>,
}

impl DependencyGraph {
    pub fn new() -> Self {
        Self {
            nodes: HashMap::new(),
        }
    }

    pub fn add_node(&mut self, name: String, range: String) -> &mut DependencyNode {
        let key = format!("{}@{}", name, range);
        self.nodes
            .entry(key.clone())
            .or_insert_with(|| DependencyNode {
                name,
                range,
                resolved: None,
                dependencies: HashMap::new(),
                dependents: HashSet::new(),
            })
    }

    pub fn get_node(&self, name: &str, version: &semver::Version) -> Option<&DependencyNode> {
        self.nodes.get(&format!("{}@{}", name, version))
    }

    pub fn detect_conflicts(&self) -> Vec<Conflict> {
        let mut by_name: HashMap<String, Vec<&DependencyNode>> = HashMap::new();
        for node in self.nodes.values() {
            by_name
                .entry(node.name.clone())
                .or_default()
                .push(node);
        }

        let mut conflicts = Vec::new();
        for (name, nodes) in by_name {
            let mut versions = Vec::new();
            for node in nodes {
                if let Some(resolved) = &node.resolved {
                    versions.push((resolved.version.clone(), node.dependents.clone()));
                }
            }
            if versions.len() > 1 {
                versions.sort_by(|a, b| b.0.cmp(&a.0));
                let highest = versions[0].0.clone();
                let compatible: Vec<_> = versions
                    .iter()
                    .filter(|(v, _)| semver::VersionReq::parse(&format!("^{}", highest)).unwrap().matches(v))
                    .collect();
                if compatible.len() != versions.len() {
                    conflicts.push(Conflict {
                        name,
                        versions: versions
                            .into_iter()
                            .map(|(v, r)| VersionConflict {
                                version: v,
                                required_by: r.into_iter().collect(),
                            })
                            .collect(),
                    });
                }
            }
        }
        conflicts
    }

    pub fn get_resolution_order(&self) -> Vec<&DependencyNode> {
        let mut visited = HashSet::new();
        let mut order = Vec::new();

        fn visit<'a>(
            node: &'a DependencyNode,
            nodes: &HashMap<String, DependencyNode>,
            visited: &mut HashSet<String>,
            order: &mut Vec<&'a DependencyNode>,
        ) {
            let key = node.key();
            if visited.contains(&key) {
                return;
            }
            visited.insert(key);
            for dep in node.dependencies.values() {
                visit(dep, nodes, visited, order);
            }
            order.push(node);
        }

        for node in self.nodes.values() {
            visit(node, &self.nodes, &mut visited, &mut order);
        }
        order
    }
}

impl Default for DependencyGraph {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone)]
pub struct Conflict {
    pub name: String,
    pub versions: Vec<VersionConflict>,
}

#[derive(Debug, Clone)]
pub struct VersionConflict {
    pub version: semver::Version,
    pub required_by: Vec<String>,
}